include/xtensor/xtiny.hpp

```diff
@@ -0,0 +1,1925 @@
+        using reference = typename base_type::reference;
+        using const_reference = typename base_type::const_reference;
+        using pointer = typename base_type::pointer;
+        using const_pointer = typename base_type::const_pointer;
+        using iterator = typename base_type::iterator;
+        using const_iterator = typename base_type::const_iterator;
+        using reverse_iterator = typename base_type::reverse_iterator;
+        using const_reverse_iterator = typename base_type::const_reverse_iterator;
+        using size_type = typename base_type::size_type;
+        using difference_type = typename base_type::difference_type;
+
+        using base_type::owns_memory;
+        using base_type::has_fixed_size;
+        using base_type::static_size;
+
+        using base_type::base_type;
+
+        xtiny();
+        xtiny(xtiny const & rhs);
+        xtiny(xtiny && rhs);
+
+        template <class T, index_t M, class R>
+        explicit xtiny(xtiny<T, M, R> const & rhs);
+
+        template <class T, class A>
+        explicit xtiny(std::vector<T, A> const & v);
+
+        template <class T, std::size_t M>
+        explicit xtiny(std::array<T, M> const & v);
+
+        xtiny & operator=(xtiny const & rhs);
+        xtiny & operator=(xtiny && rhs);
+
+        xtiny & operator=(value_type const & v);
+
+        template <class T, class A>
+        xtiny & operator=(std::vector<T, A> const & v);
+
+        template <class T, std::size_t M>
+        xtiny & operator=(std::array<T, M> const & v);
+
+        template <class U, index_t M, class R>
+        xtiny & operator=(xtiny<U, M, R> const & rhs);
+
+        using base_type::assign;
+
+        void assign(std::initializer_list<value_type> v);
+
+        using base_type::data;
+
+        using base_type::operator[];
+        reference at(size_type i);
+        constexpr const_reference at(size_type i) const;
+
+        reference front();
+        reference back();
+        constexpr const_reference front() const;
+        constexpr const_reference back()  const;
+
+        template <index_t FROM, index_t TO>
+        auto subarray();
+        template <index_t FROM, index_t TO>
+        auto subarray() const;
+        auto subarray(size_type FROM, size_type TO);
+        auto subarray(size_type FROM, size_type TO) const;
+
+        auto erase(size_type m) const;
+        auto pop_front() const;
+        auto pop_back() const;
+
+        auto insert(size_type m, value_type v) const;
+        auto push_front(value_type v) const;
+        auto push_back(value_type v) const;
+
+        using base_type::begin;
+        constexpr const_iterator cbegin() const;
+        iterator end();
+        constexpr const_iterator end() const;
+        constexpr const_iterator cend() const;
+
+        using base_type::rbegin;
+        constexpr const_reverse_iterator crbegin() const;
+        reverse_iterator rend();
+        constexpr const_reverse_iterator rend() const;
```

Comment:
I tried this, but I am quite confident it doesn't compile with C++14 (constexpr reverse iterator will be a C++17 feature).

Reply:
It worked on all platforms covered in your test suite. (The current red light seems to be due to failing setup code in the gcc 6 environment -- the test doesn't even get to the point of compilation.)

Reply:
yep. because constexpr variables are probably not covered by the test. 

```
constexpr xtiny<...> test({1,2,3});
constexpr auto it = test.rbegin();
```

will likely not compile.

Reply:
anyways, for the constexpr to have any effect we probably need to make some constructors constexpr as well ... 

I accidentally found an interesting example of a hybrid constexpr class today. Might be a source for some inspiration for potential future improvements :)

https://github.com/akrzemi1/Optional/blob/master/optional.hpp#L319-L346

Reply:
This already fails in the constructor line. I gues, when the constructor goes through, `rbegin()` will compile as well.

Reply:
I mainly defined `constexpr ... rbegin()` to potentially accommodate your `fixed_vector`.

Reply:
I'll make some experiments today to see how constexpr and base classes can be combined. Maybe I'll have a proposal ready tonight.

Reply:
So far I can't get anything nice to work. I'm unsure if it's possible to get a `constexpr std::array-like` type to work ... the problem is that aggregate initialization with initializer list doesn't work when inheriting (as it's not an aggregate anymore).

Reply:
I suggest to try the `fixed_vector` adapter I described here: https://github.com/QuantStack/xtensor/pull/591#issuecomment-352371871. Then you only need a default constructor.